### ISSUE TITLE
SSLError CERTIFICATE_VERIFY_FAILED for sampledata.download

### ISSUE THREAD

**Issue by cdeil** [bokeh/bokeh#2310]
I have bokeh 0.9.0 via Macports and wanted to download the sample data as described here:
http://bokeh.pydata.org/en/latest/docs/quickstart.html#sample-data

But I'm getting this SSL CERTIFICATE_VERIFY_FAILED error on Python 2 and 3:
```
$ python -c "import bokeh.sampledata; bokeh.sampledata.download()"
Creating /Users/deil/.bokeh/data directory
Using data directory: /Users/deil/.bokeh/data
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 1182, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/http/client.py", line 1088, in request
    self._send_request(method, url, body, headers)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/http/client.py", line 1126, in _send_request
    self.endheaders(body)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/http/client.py", line 1084, in endheaders
    self._send_output(message_body)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/http/client.py", line 922, in _send_output
    self.send(msg)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/http/client.py", line 857, in send
    self.connect()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/http/client.py", line 1231, in connect
    server_hostname=server_hostname)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/ssl.py", line 365, in wrap_socket
    _context=self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/ssl.py", line 583, in __init__
    self.do_handshake()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/ssl.py", line 810, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bokeh/sampledata/__init__.py", line 72, in download
    _getfile(base_url, file_name, data_dir, progress=progress)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bokeh/sampledata/__init__.py", line 78, in _getfile
    url = urlopen(file_url)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 161, in urlopen
    return opener.open(url, data, timeout)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 463, in open
    response = self._open(req, data)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 481, in _open
    '_open', req)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 441, in _call_chain
    result = func(*args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 1225, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/urllib/request.py", line 1184, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)>
```
```
$ python2.7 -c "import bokeh.sampledata; bokeh.sampledata.download()"
Using data directory: /Users/deil/.bokeh/data
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/deil/Library/Python/2.7/lib/python/site-packages/bokeh/sampledata/__init__.py", line 71, in download
    _getfile(base_url, file_name, data_dir, progress=progress)
  File "/Users/deil/Library/Python/2.7/lib/python/site-packages/bokeh/sampledata/__init__.py", line 75, in _getfile
    u = urlopen(url)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 431, in open
    response = self._open(req, data)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 449, in _open
    '_open', req)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1240, in https_open
    context=self._context)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1197, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)>
```

**Comment by damianavila** [comment]
Which pythons version are you using? I think it is related with this: https://bugs.python.org/issue22417
Check this too: http://stackoverflow.com/a/30209998/2785603

**Comment by cdeil** [comment]
I'm using Python 3.4.3 and 2.7.9 with latest Macports (2.3.3) on latest OS X (10.10.3).

If it's just me and unclear how to fix this on your or my end without extra work, feel free to close this issue. I just wanted to try out bokeh a bit, I don't really need the example files.

**Comment by bryevdv** [comment]
Have not heard any other instances of this. If anyone encounters the same or a similar problem, please re-open.

**Comment by ciacicode** [comment]
Just had this issue on my ubuntu 16.04, using a conda environment with Python 2.7

`$ bokeh sample data`
`ERROR:<urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)>`

**Comment by bryevdv** [comment]
@ciacicode I just tried this and was not able to reproduce, the data comes straight off an S3 bucket, so I think this must be some kind of local configuration issue. Perhaps you can provide more details about your system, versions (including ssl lib), etc.

**Comment by ivanfrain** [comment]
Hi all, hi bryevdv,

I am also running into this problem. Below more information about my problem. After looking for a solution around the internet, I was not able to resolve my problem.

I am running alpine linux 3.4 in a docker container with following software installed : 

* Python 3.5.2
* OpenSSL 1.0.2j  26 Sep 2016
* pip 8.1.2 from /usr/lib/python3.5/site-packages (python 3.5)
* pip3 list = 
  - bokeh (0.12.3)
  - certifi (2016.9.26)
  - cffi (1.8.3)
  - cryptography (1.5.2)
  - Cython (0.25.1)
  - decorator (4.0.10)
  - entrypoints (0.2.2)
  - idna (2.1)
  - ipykernel (4.5.0)
  - ipython (5.1.0)
  - ipython-genutils (0.1.0)
  - ipywidgets (5.2.2)
  - Jinja2 (2.8)
  - jsonschema (2.5.1)
  - jupyter (1.0.0)
  - jupyter-client (4.4.0)
  - jupyter-console (5.0.0)
  - jupyter-core (4.2.0)
  - MarkupSafe (0.23)
  - mistune (0.7.3)
  - nbconvert (4.2.0)
  - nbformat (4.1.0)
  - notebook (4.2.3)
  - numpy (1.11.2)
  - pandas (0.19.0)
  - pexpect (4.2.1)
  - pickleshare (0.7.4)
  - pip (8.1.2)
  - prompt-toolkit (1.0.8)
  - ptyprocess (0.5.1)
  - pyasn1 (0.1.9)
  - pycparser (2.17)
  - Pygments (2.1.3)
  - pyOpenSSL (16.2.0)
  - python-dateutil (2.5.3)
  - pytz (2016.7)
  - PyYAML (3.12)
  - pyzmq (16.0.0)
  - qtconsole (4.2.1)
  - requests (2.11.1)
  - setuptools (20.10.1)
  - simplegeneric (0.8.1)
  - six (1.10.0)
  - terminado (0.6)
  - tornado (4.4.2)
  - traitlets (4.3.1)
  - wcwidth (0.1.7)
  - widgetsnbextension (1.2.6)

**Running bokeh sampledata from the command line**

```
718e90408818:~$ bokeh sampledata
Using data directory: /home/nbuser/.bokeh/data
ERROR:<urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed
(_ssl.c:645)>

```
**Running bokeh from a jupyter notebook**

```
import bokeh.sampledata;
bokeh.sampledata.download();

Using data directory: /home/nbuser/.bokeh/data
---------------------------------------------------------------------------
SSLError                                  Traceback (most recent call last)
/usr/lib/python3.5/urllib/request.py in do_open(self, http_class, req, **http_conn_args)
   1253             try:
-> 1254                 h.request(req.get_method(), req.selector, req.data, headers)
   1255             except OSError as err: # timeout error

/usr/lib/python3.5/http/client.py in request(self, method, url, body, headers)
   1105         """Send a complete request to the server."""
-> 1106         self._send_request(method, url, body, headers)
   1107 

/usr/lib/python3.5/http/client.py in _send_request(self, method, url, body, headers)
   1150             body = _encode(body, 'body')
-> 1151         self.endheaders(body)
   1152 

/usr/lib/python3.5/http/client.py in endheaders(self, message_body)
   1101             raise CannotSendHeader()
-> 1102         self._send_output(message_body)
   1103 

/usr/lib/python3.5/http/client.py in _send_output(self, message_body)
    933 
--> 934         self.send(msg)
    935         if message_body is not None:

/usr/lib/python3.5/http/client.py in send(self, data)
    876             if self.auto_open:
--> 877                 self.connect()
    878             else:

/usr/lib/python3.5/http/client.py in connect(self)
   1259             self.sock = self._context.wrap_socket(self.sock,
-> 1260                                                   server_hostname=server_hostname)
   1261             if not self._context.check_hostname and self._check_hostname:

/usr/lib/python3.5/ssl.py in wrap_socket(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname)
    376                          server_hostname=server_hostname,
--> 377                          _context=self)
    378 

/usr/lib/python3.5/ssl.py in __init__(self, sock, keyfile, certfile, server_side, cert_reqs, ssl_version, ca_certs, do_handshake_on_connect, family, type, proto, fileno, suppress_ragged_eofs, npn_protocols, ciphers, server_hostname, _context)
    751                         raise ValueError("do_handshake_on_connect should not be specified for non-blocking sockets")
--> 752                     self.do_handshake()
    753 

/usr/lib/python3.5/ssl.py in do_handshake(self, block)
    987                 self.settimeout(None)
--> 988             self._sslobj.do_handshake()
    989         finally:

/usr/lib/python3.5/ssl.py in do_handshake(self)
    632         """Start the SSL/TLS handshake."""
--> 633         self._sslobj.do_handshake()
    634         if self.context.check_hostname:

SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)

During handling of the above exception, another exception occurred:

URLError                                  Traceback (most recent call last)
<ipython-input-1-1673400b2a17> in <module>()
      1 import bokeh.sampledata;
      2 
----> 3 bokeh.sampledata.download();

/usr/lib/python3.5/site-packages/bokeh/sampledata/__init__.py in download(progress)
     79 
     80     for base_url, file_name in files:
---> 81         _getfile(base_url, file_name, data_dir, progress=progress)
     82 
     83 def _getfile(base_url, file_name, data_dir, progress=True):

/usr/lib/python3.5/site-packages/bokeh/sampledata/__init__.py in _getfile(base_url, file_name, data_dir, progress)
     85     file_path = join(data_dir, file_name)
     86 
---> 87     url = urlopen(file_url)
     88 
     89     with open(file_path, 'wb') as file:

/usr/lib/python3.5/urllib/request.py in urlopen(url, data, timeout, cafile, capath, cadefault, context)
    161     else:
    162         opener = _opener
--> 163     return opener.open(url, data, timeout)
    164 
    165 def install_opener(opener):

/usr/lib/python3.5/urllib/request.py in open(self, fullurl, data, timeout)
    464             req = meth(req)
    465 
--> 466         response = self._open(req, data)
    467 
    468         # post-process response

/usr/lib/python3.5/urllib/request.py in _open(self, req, data)
    482         protocol = req.type
    483         result = self._call_chain(self.handle_open, protocol, protocol +
--> 484                                   '_open', req)
    485         if result:
    486             return result

/usr/lib/python3.5/urllib/request.py in _call_chain(self, chain, kind, meth_name, *args)
    442         for handler in handlers:
    443             func = getattr(handler, meth_name)
--> 444             result = func(*args)
    445             if result is not None:
    446                 return result

/usr/lib/python3.5/urllib/request.py in https_open(self, req)
   1295         def https_open(self, req):
   1296             return self.do_open(http.client.HTTPSConnection, req,
-> 1297                 context=self._context, check_hostname=self._check_hostname)
   1298 
   1299         https_request = AbstractHTTPHandler.do_request_

/usr/lib/python3.5/urllib/request.py in do_open(self, http_class, req, **http_conn_args)
   1254                 h.request(req.get_method(), req.selector, req.data, headers)
   1255             except OSError as err: # timeout error
-> 1256                 raise URLError(err)
   1257             r = h.getresponse()
   1258         except:

URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)>

```

Seems it is something to deal with the docker container hostname or name resolution inside the SSL resolution process but I was not able to figure out what.

Any ideas ?

NB: I was able to run the bookeh sampledata command succesfully from the host (the host of docker container). Thus it might be a misconfiguration of the docker container itself.

Thanks for your help.

**Comment by birdsarah** [comment]
@ivanfrain, can I suggest starting by posting your question on the bokeh mailing list https://groups.google.com/a/continuum.io/forum/#!forum/bokeh. This is a closed issue, so not the best place to get support.

**Comment by ivanfrain** [comment]
@birdsarah thank you for your answer. I have sent the question to the bokeh group : https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/98kME6JAE0Y

**Comment by Denisevi4** [comment]
I was able to solve this by executing the following:

import ssl
ssl._create_default_https_context = ssl._create_unverified_context

**Comment by guyts** [comment]
I have just installed the two using conda, and am experiencing the issues as mentioned above.
My installation was simply `conda install -c pyviz holoviews bokeh` as specified on [HV docs](http://holoviews.org/index.html).
When running `hv.extension('bokeh')` I'm getting the following error:

```
WARNING:param.notebook_extension: Holoviews bokeh extension could not be imported, it raised the following exception: SSLError('unknown error (_ssl.c:2994)')
Traceback (most recent call last):
..
ImportError: None of the backends could be imported
```
I tried @Denisevi4 method but it didn't work in my case.
Running bokeh 1.0.4 and hv 1.11.0 on Windows 10, Py3.7 (Anaconda 2019.03).
